### PR TITLE
support issuing vote instructions from system account

### LINKED ISSUE #4173: vote instructions are not issued from a system account
#### Problem
vote instruction constructors need a "from" parameter that needs to be a system account to pay the transaction fees

#### Proposed Solution
add a "from" to all instruction constructors, a la stake_instruction

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 vote instruction constructors need a "from" parameter that needs to be a system account to pay the transaction fees

 #### Summary of Changes
 update vote_instruction and customers to issue from a system account

Fixes #4173

### CODE REVIEW

**Comment by garious** [comment]
Why rip out use of the KeypairUtil trait?

**Comment by rob-solana** [comment]
> Why rip out use of the KeypairUtil trait?

I was fighting Transaction::sign(), which wouldn't take a Keypair and a KeypairUtil together in a slice.  In order to get this closed out, I took the simple path.  I can look at bringing it back now, either with Transaction::partial_sign() or with a rust trick.

File: core/src/replay_stage.rs [hunk 1]
```diff
@@ -322,13 +322,20 @@ impl ReplayStage {
...
+                &node_keypair.pubkey(),
+                &vote_account,
+                &voting_keypair.pubkey(),
+                locktower.recent_votes(),
             );
+
+            let mut vote_tx = Transaction::new_unsigned_instructions(vec![vote_ix]);
+            let blockhash = bank.last_blockhash();
+            vote_tx.partial_sign(&[node_keypair.as_ref()], blockhash);
+            vote_tx.partial_sign(&[voting_keypair.as_ref()], blockhash);
```
Comment (rob-solana) [inline core/src/replay_stage.rs]: @garious, rust can't handle `&[node_keypair.as_ref(), voting_keypair.as_ref()]`

Comment (garious) [reply core/src/replay_stage.rs]: That's probably only true if they are different implementations of KeypairUtil.  Your workaround of making two calls is a very good one. The alternative is to change the APIs to take trait objects, but that a malloc to box every keypair reference (which I'm okay with).

**Comment by garious** [comment]
Do you have a way to test the new wallet code? So much indirection. This stuff is hard to grok!

File: core/src/replay_stage.rs (hunk 1 above)
Comment (rob-solana) [reply core/src/replay_stage.rs]: > That's probably only true if they are different implementations of KeypairUtil.
nope.  both node_keypair and voting_keypair come from the same place...

@sagar-solana and I looked at it and found some really weird stuff going on with rust

**Comment by rob-solana** [comment]
> Do you have a way to test the new wallet code? So much indirection. This stuff is hard to grok!

there are a bunch of tests in there...  can you clarify?

**Comment by garious** [comment]
Something that demonstrates the need for different first and third pubkeys. Looks like you pass the identical ones into both parameters. Maybe I'm just struggling with the diff.

**Comment by rob-solana** [comment]
the new wallet code follows directly from vote_instruction stuff.  first keypair is the issuer of the TX, needs to pay TX fees.  the 2nd keypair is the authorized vote signer, which we currently default everywhere to the vote account, but we support them being different.

I wish instruction stuff and transaction stuff were more closely associated...  today, there's no way for vote_instruction to indicate who needs to sign the transaction that would be made.  the ordering of pubkeys and keypairs is also super error-prone, too.  only the person who calls Transaction::new_[un]signed_instructions() can know (welll they have to infer it) in what order the keys have to be presented.

**Comment by rob-solana** [comment]
a test that fails in wallet when the wrong parameters were presented would have to run a full bank with the vote program included (I think).

**Pull request merged**